File: GOF/Builder/java/src/Maze/Maze.java
package Maze;

import java.util.LinkedList;

public class Maze {

	private LinkedList<Room> m_rooms = new LinkedList<Room>();

	public Room roomNo(int n) {
		for(Room room : m_rooms) {
			if (room.no() == n)
				return room;
		}
		return null;
	}

	public void addRoom(Room room) {
		m_rooms.addLast(room);
	}
}

File: GOF/Builder/cpp/maze/Maze.cpp
#include "Maze.h"
#include "Room.h"
#include <algorithm>

std::weak_ptr<Room> Maze::roomNo(int n)
{
	auto room = std::find_if(m_rooms.begin(), m_rooms.end(),
			[n](auto room) { return room->no() == n; });
	return room == m_rooms.end() ? nullptr : *room;
}

void Maze::addRoom(std::shared_ptr<Room> room)
{
	m_rooms.push_back(room);
}
